[Local Catalog] Show items from GRDB in POS

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

Merge after: #16231

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR starts using the observable GRDB data source added in #16231 to show items from GRDB in the item list.

Note that there's nothing to enforce that GRDB actually has items in it at the moment... so you can definitely break this, but those guards are covered by other tickets and this is feature flagged.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

### Check we're not breaking production...
- Set the `pointOfSaleLocalCatalogi1` feature flag to false
- Open the POS
- Observe that browsing and scrolling works as expected, and use WormHoly or a proxy tool to verify that products are loaded from the network.

### Test the GRDB item controller
- Reset the feature flag
- Launch the app
- Wait for console log messages to confirm the catalog sync status for your store
- Open POS
- Observe that products are shown, and that you can open variation lists
- Scroll down, observe that subsequent pages are loaded
- Turn on airplane mode, observe that you can still fully browse the catalog.

Example sync log messages:
```
📋 POSCatalogSyncCoordinator: Site 216335538 has catalog size 91, with 45 products and 46 variations
📋 POSCatalogSyncCoordinator: Last sync for site 216335538 was 16344s ago (max: 86400s), sync not needed
```

### Optional: test with a larger site
- Change the value for `POSCatalogSyncCoordinator.Constants.defaultSizeLimitForPOSCatalog` to 100000
- Run the app and pick a store with a large catalog
- Wait for the sync to complete (it will take some time and gives no progress info before completion)
- Open POS
- Observe that you can interact with the full catalog

Note that with larger catalogs, you start to hit scrolling performance issues. These are outside the scope of current work, but we may want to address them before too long. I don't think it's down to the mapping happening on the main thread, and it may even be that we're hitting some SwiftUI limitations.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: https://woomobilep2.wordpress.com/2024/05/06/woocommerce-mobile-quality-report-march-april/#comment-12036 -->

Note that Search and Barcode Scanning are still using the network-based controller, so have delays.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/89642075-b4f4-4cf3-8417-c1608a982922


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

joshheald committed:
Add .initial state to ItemListState

Adds a new `.initial` case to `ItemListState` to represent "not yet loaded" state, distinct from "actively loading".

Benefits:
- More explicit state machine: initial → loading → loaded/empty/error
- Foundation for delayed loading indicators (300ms threshold)
- Better UX for instant GRDB loads (no flashing skeleton)

Updated existing `PointOfSaleItemsController` to:
- Start with `.initial` instead of `.loading([])`
- Transition to `.loading([])` on first load
- Preserve existing behavior for refreshes

joshheald committed:
Add PointOfSaleObservableItemsController

Creates a new controller that wraps `POSObservableDataSourceProtocol` for GRDB-based item loading.

Key features:
- Conforms to `PointOfSaleItemsControllerProtocol` (same interface as existing controller)
- Computed `itemsViewState` from observable data source properties
- Automatic UI updates via observation (no manual state management)
- Supports products and variations with `.initial` state
- Tracks current parent for variation state mapping

Implementation notes:
- No pagination tracking needed (data source handles it)
- No search support (out of scope for this PR)

joshheald committed:
Add tests for PointOfSaleObservableItemsController

Comprehensive test coverage for the new observable controller:

- Initial state (loading container, initial root)
- Product loading (loaded, empty, error states)
- Variation loading (loaded, empty states)
- Pagination (hasMoreItems flags)
- Independence of products and variations
- Parent switching behavior
- Error handling
- Loading state preservation

Fix tests to use proper POSItem construction

Use helper methods to create POSSimpleProduct and POSVariation structs
instead of non-existent .fake() method, matching the pattern used in
MockPOSItemProvider.

Changes:
- Added makeSimpleProduct() helper
- Added makeVariation() helper
- Updated all 12 tests to use proper construction

joshheald committed:
Use PointOfSaleObservableItemsController when GRDB is available

When grdbManager, catalogSyncCoordinator, and the pointOfSaleLocalCatalogi1
feature flag are all present, use the new PointOfSaleObservableItemsController
backed by GRDBObservableDataSource instead of the standard
PointOfSaleItemsController.

This enables reactive database-driven item lists for stores with local
catalog sync enabled.

joshheald commented:
@staskus Sorry it took longer than I thought it would, but here's the second PR as mentioned this morning.

I'm AFK next week – happy for you to merge them then if you prefer not to review this afternoon. As long as there's nothing major needed as changes!

joshheald commented on Modules/Sources/PointOfSale/Models/ItemListState.swift:
```diff
@@ -1,6 +1,7 @@
 import enum Yosemite.POSItem
 
 enum ItemListState {
+    case initial
```
I added this state for more accurate modelling of what's happening.

My intention is to delay showing the loading indicators when we open variations with GRDB – it causes a slight flicker at the moment. This is some groundwork for that.

https://linear.app/a8c/issue/WOOMOB-1494/woo-poslocal-catalog-add-a-delay-before-showing-loading-indicators

joshheald committed:
Periphery ignore

joshheald committed:
Fix periphery ignore

joshheald commented:
@staskus I also tested that it observes changes: [this branch adds an incremental sync button to the ... menu](https://github.com/woocommerce/woocommerce-ios/compare/issue/woomob-1088-observable-itemcontroller-for-grdb...issue/woomob-1088-observable-itemcontroller-for-grdb-testing). I used that by editing products on the web and then syncing.

Trashing a product didn't remove it, which is probably down to how we request them or filter the products from the db... but I'll fix it in a separate PR. https://linear.app/a8c/issue/WOOMOB-1493/woo-poslocal-catalog-ensure-that-trashed-products-are-updated-to-not

Other changes and additions worked as expected.

staskus commented on Modules/Sources/PointOfSale/Controllers/PointOfSaleObservableItemsController.swift:
```diff
@@ -0,0 +1,160 @@
+    func loadItems(base: ItemListBaseItem) async {
+        switch base {
+        case .root:
+            hasLoadedProducts = true
```
Should we delay it until the items are loaded? Same with other similar flags.

joshheald replied:
Done in 2bdf4ab

staskus commented on Modules/Sources/PointOfSale/Controllers/PointOfSaleObservableItemsController.swift:
```diff
@@ -0,0 +1,160 @@
+    func refreshItems(base: ItemListBaseItem) async {
+        switch base {
+        case .root:
+            dataSource.refresh()
```
So, this is not implemented. Is the idea to trigger an incremental sync when refresh is called?

joshheald replied:
Yes, that's right

staskus commented on Modules/Sources/PointOfSale/Controllers/PointOfSaleObservableItemsController.swift:
```diff
@@ -0,0 +1,160 @@
+        }
+
+        // Error state
+        if let error = dataSource.error, items.isEmpty {
```
What kind of errors can we expect here? Since this comes from database, only something super-unexpected I suppose?

joshheald replied:
It should be extremely rare – these are really simple select queries...

- Database corruption
- Programming errors during development, e.g. relationships missing, or incorrect query setup – but those shouldn't ever happen in production

Errors are more likely to appear when we sync the data, because we're relying on it coming down with the correct schema and logically consistent. These will happen when we write stuff to the DB, and will prevent the whole write transacion if they're not handled in other code... so they won't get as far as being shown here. It's primarily included for completeness with the existing code, and to make it easier to spot bugs in dev.

I noticed that we only have one error property, shared between variations and products, which could result in an error state being shown for the wrong screen. I've separated them out. 46b634a

staskus reviewed (commented):
Thank you, @joshheald, for work! It took some time to review and understand. I'm not yet that familiar with `GRDB`, plus there's complexity in clean state management and observability. We don't want the views to stuck in some "no man's land" state.

I think it's mostly good. The comments are in both PRs. I tested with smaller and larger stores. I think there are a couple of things that need to be addressed. I don't see much value in pushing this to trunk without addressing and understanding them properly. I will see if I find time to come back to this after finishing the synchronization work. Otherwise, we can discuss next week and fix things or move to other tasks and merge this work.

staskus reviewed (approved):
I retested on Friday and today as well comparing the new version with the previous version and all the same products, including variable ones are loaded. The pagination works, including with large datasets. 

As you mentioned, there are still ways to make it solid, and handle the cases where the catalog is not loaded yet but otherwise it works well and is super fast 👍

joshheald committed:
Separate product and variation errors to prevent context leakage

Previously, a single error property was shared between products and
variations, which could cause errors to leak between the root and
variation parent contexts.

Now using separate productError and variationError properties so that:
- Product loading errors only affect the root state
- Variation loading errors only affect the parent item state
- Errors don't interfere with each other

Updated:
- POSObservableDataSourceProtocol
- GRDBObservableDataSource implementation
- PointOfSaleObservableItemsController state computation
- MockPOSObservableDataSource for tests

joshheald committed:
Clear variation error when switching to new parent

When loading variations for a new parent product, clear any previous
variation error to prevent it from leaking to the new parent's context.

This ensures each parent starts with a clean error state, matching the
behavior of clearing variationItems.

joshheald committed:
Set `hasLoaded` after we actually load

joshheald committed:
Fix tests

joshheald committed:
Fix test
